[CIR] Add constant record ILE support

mmha opened the pull request:
This patch adds basic support for constant record initializer list expressions. There's a couple of limitations:

* No zero initialized padding bytes in C mode
* No bitfields
* No designated initializer lists
* Record alignments are not calculated, yet
* ILEs of derived records don't work, yet
* The constant attribute is not propagated to the backend, resulting in non-constants being emitted in the LLVM IR

mmha committed:
clang-format

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuilder.cpp:
```diff
@@ -130,6 +132,44 @@ void CIRGenBuilderTy::computeGlobalViewIndicesFromFlatOffset(
+  return mlir::cast<mlir::TypedAttr>(attr).getType();
+}
+
+cir::RecordType clang::CIRGen::CIRGenBuilderTy::getCompleteRecordType(
```
I don't love having both `getCompleteRecordTy` and `getCompleteRecordType`. Can we give them the same name and put the common parameters in the same order?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuilder.cpp:
```diff
@@ -130,6 +132,44 @@ void CIRGenBuilderTy::computeGlobalViewIndicesFromFlatOffset(
+
+  // Record type not specified: create anon record type from members.
+  if (!recordTy) {
+    llvm::SmallVector<mlir::Type, 8> members;
```
Why not call `getCompleteRecordTy` here?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,571 @@
+    return true;
+  }
+
+  // Uncommon case: constant overlaps what we've already created.
```
Maybe refactor all of the "uncommon" handling into a separate PR?

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -283,6 +283,16 @@ Type RecordType::getLargestMember(const ::mlir::DataLayout &dataLayout) const {
+  if (getPacked() != other.getPacked())
+    return false;
+
+  return getMembers() == other.getMembers();
```
The bases are represented as members in the record.

mmha committed:
Address review feedback

mmha committed:
Add more tests

mmha committed:
Add simple test case

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+
+    // Accumulate and sort bases, in order to visit them in address order, which
+    // may not be the same as declaration order.
+    SmallVector<BaseInfo, 8> bases;
```
```suggestion
    SmallVector<BaseInfo> bases;
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+      CharUnits baseOffset = layout.getBaseClassOffset(bd);
+      bases.push_back(BaseInfo(bd, baseOffset, index));
+    }
+    llvm::stable_sort(bases);
```
I'd be a bit surprised if these aren't already correctly sorted. Have you seen otherwise?

mmha replied:
I thought that we might need this for empty bases classes but that doesn't appear to be the case. I then replaced this `stable_sort` call in classic codegen with an `assert(is_sorted(bases))` and the tests still pass.

To me it's not obvious that the bases are always added in the correct order so I'll replace this `stable_sort` with that assertion wrapped in an `#ifdef EXPENSIVE_CHECKS` block.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+    SmallVector<BaseInfo, 8> bases;
+    bases.reserve(cd->getNumBases());
+    for (auto [index, base] : llvm::enumerate(cd->bases())) {
+      assert(!base.isVirtual() && "should not have virtual bases here");
```
Where is this prevented?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+                               CharUnits offset) {
+  const ASTRecordLayout &layout = cgm.getASTContext().getASTRecordLayout(rd);
+  if (const CXXRecordDecl *cd = dyn_cast<CXXRecordDecl>(rd)) {
+    assert(!cir::MissingFeatures::vtableInitialization());
```
This seems more like it should be an `errorNYI` if the class requires vtable initialization.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -903,8 +1440,7 @@ mlir::Attribute ConstantEmitter::tryEmitPrivate(const APValue &value,
   case APValue::Union:
-    cgm.errorNYI("ConstExprEmitter::tryEmitPrivate struct or union");
-    return {};
+    return ConstRecordBuilder::buildRecord(*this, value, destType);
```
Which case gets us here? Is this covered by one of the tests?

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+  bool padded = false;
+  ArrayRef<mlir::TypedAttr> unpackedElems = elems;
+
+  llvm::SmallVector<mlir::TypedAttr, 32> unpackedElemStorage;
```
There is a commit from 2012 that bumped it: https://github.com/llvm/llvm-project/commit/7ac747245f1c459c6257fa163999d29fe4dd423a

I'd say let's keep the explicit size here

mmha committed:
[CIR] Add constant record ILE support

This patch adds basic support for constant record initializer list expressions. There's a couple of limitations:

* No zero initialized padding bytes in C mode
* No bitfields
* No designated initializer lists
* Record alignments are not calculated, yet
* ILEs of derived records don't work, yet
* The constant attribute is not propagated to the backend, resulting in non-constants being emitted in the LLVM IR

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp:
```diff
@@ -21,20 +20,526 @@
+
+  CharUnits getAlignment(const mlir::TypedAttr c) const {
+    return CharUnits::fromQuantity(
+        dataLayout.getAlignment(c.getType(), /*abiOrPref=*/true));
```
The function definition uses `useABIAlign` but the declaration still uses the old name I fixed that up, too.

mmha committed:
Address review feedback

andykaylor reviewed (approved):
lgtm
